In [3]:
import numpy as np

# Урок 3

# Матрицы и матричные операции. Часть 1

__1.__ Установить, какие произведения матриц $AB$ и $BA$ определены, и найти размерности полученных матриц:

   а) $A$ — матрица $4\times 2$, $B$ — матрица $4\times 2$;
    
   б) $A$ — матрица $2\times 5$, $B$ — матрица $5\times 3$;
    
   в) $A$ — матрица $8\times 3$, $B$ — матрица $3\times 8$;
    
   г) $A$ — квадратная матрица $4\times 4$, $B$ — квадратная матрица $4\times 4$.

а) Произведения матриц AB и BA не определены

б) Произведение AB определено, размерность полученной матрицы 2 х 3. Произведение BA не определено

в) Произвеление AB определено, размерность полученной матрицы 8 х 8. Произведение BA определено, размерность полученной матрицы 3 х 3.

г) Произвеления AB и BA определены, размерность полученных матриц 4 х 4

__2.__ Найти сумму и произведение матриц $A=\begin{pmatrix}
1 & -2\\ 
3 & 0
\end{pmatrix}$ и $B=\begin{pmatrix}
4 & -1\\ 
0 & 5
\end{pmatrix}.$

$$\begin{pmatrix}
1 & -2\\ 
3 & 0
\end{pmatrix} + \begin{pmatrix}
4 & -1\\ 
0 & 5
\end{pmatrix}=
\begin{pmatrix}
5 & -3\\ 
3 & 5
\end{pmatrix}$$

In [7]:
a = np.array([[1, -2], [3, 0]])
b = np.array([[4, -1], [0, 5]])

print(f'Матрица A\n{a}\n')
print(f'Матрица B\n{b}\n')
print(f'Матрица С = A + B\n{a + b}')

Матрица A
[[ 1 -2]
 [ 3  0]]

Матрица B
[[ 4 -1]
 [ 0  5]]

Матрица С = A + B
[[ 5 -3]
 [ 3  5]]


$$\begin{pmatrix}
1 & -2\\ 
3 & 0
\end{pmatrix} \begin{pmatrix}
4 & -1\\ 
0 & 5
\end{pmatrix}=\begin{pmatrix}
(1\cdot4 + (-2)\cdot0) & (1\cdot(-1) + (-2)\cdot5)\\ 
(3\cdot4 + 0\cdot0) & (3\cdot(-1) + 0\cdot5)
\end{pmatrix}=
\begin{pmatrix}
4 & -11\\ 
12 & -3
\end{pmatrix}$$

In [6]:
A = np.array([[1, -2], [3, 0]])
B = np.array([[4, -1], [0, 5]])

print(f'Матрица A\n{A}')
print(f'Матрица B\n{B}')
print(f'Матрица AB\n{np.dot(A, B)}')

Матрица A
[[ 1 -2]
 [ 3  0]]
Матрица B
[[ 4 -1]
 [ 0  5]]
Матрица AB
[[  4 -11]
 [ 12  -3]]


__3.__ Из закономерностей сложения и умножения матриц на число можно сделать вывод, что матрицы одного размера образуют линейное пространство. Вычислить линейную комбинацию $3A-2B+4C$ для матриц $A=\begin{pmatrix}
1 & 7\\ 
3 & -6
\end{pmatrix}$, $B=\begin{pmatrix}
0 & 5\\ 
2 & -1
\end{pmatrix}$, $C=\begin{pmatrix}
2 & -4\\ 
1 & 1
\end{pmatrix}.$

$$3\cdot\begin{pmatrix}
1 & 7\\ 
3 & -6
\end{pmatrix} = 
\begin{pmatrix}
3 & 21\\ 
9 & -18
\end{pmatrix}$$

$$(-2)\cdot\begin{pmatrix}
0 & 5\\ 
2 & -1
\end{pmatrix} = 
\begin{pmatrix}
0 & -10\\ 
-4 & 2
\end{pmatrix}$$

$$4\cdot\begin{pmatrix}
2 & -4\\ 
1 & 1
\end{pmatrix} = 
\begin{pmatrix}
8 & -16\\ 
4 & 4
\end{pmatrix}$$

$$\begin{pmatrix}
3 & 21\\ 
9 & -18
\end{pmatrix} + \begin{pmatrix}
0 & -10\\ 
-4 & 2
\end{pmatrix}
 + \begin{pmatrix}
8 & -16\\ 
4 & 4
\end{pmatrix}=
\begin{pmatrix}
11 & -5\\ 
9 & -12
\end{pmatrix}$$

$$3A-2B+4C = 3 \times \begin{pmatrix}
1 & 7\\ 
3 & -6
\end{pmatrix} - 2 \times \begin{pmatrix}
0 & 5\\ 
2 & -1 
\end{pmatrix} + 4 \times \begin{pmatrix}
2 & -4\\ 
1 & 1
\end{pmatrix} = \begin{pmatrix} 
3 \cdot 1 - 2 \cdot 0 + 4 \cdot 2 & 3 \cdot 7 - 2 \cdot 5 + 4 \cdot (-4) \\
3 \cdot 3 - 2 \cdot 2 + 4 \cdot 1 & 3 \cdot (-6) - 2 \cdot (-1) + 4 \cdot 1
\end{pmatrix} = \begin{pmatrix} 
11 & -5 \\
9 & -12
\end{pmatrix}  
$$

In [14]:
A = np.array([[1, 7], [3, -6]])
B = np.array([[0, 5], [2, -1]])
C = np.array([[2, -4], [1, 1]])
k = 3
m = -2
n = 4

print(f'Матрица А\n{A}\n')
print(f'Матрица B\n{B}\n')
print(f'Матрица C\n{C}\n')
print(f'Матрица 3*А - 2*B + 4*C\n{k*A + m*B + n*C}')

Матрица А
[[ 1  7]
 [ 3 -6]]

Матрица B
[[ 0  5]
 [ 2 -1]]

Матрица C
[[ 2 -4]
 [ 1  1]]

Матрица 3*А - 2*B + 4*C
[[ 11  -5]
 [  9 -12]]


__4.__ Дана матрица $A=\begin{pmatrix}
4 & 1\\ 
5 & -2\\ 
2 & 3
\end{pmatrix}$.
Вычислить $AA^{T}$ и $A^{T}A$

$$A^{T}=\begin{pmatrix}
4 & 5 & 2 \\
1 & -2 & 3
\end{pmatrix}$$

$$A \times A^{T}=\begin{pmatrix}
4 & 1\\ 
5 & -2\\ 
2 & 3
\end{pmatrix} \times \begin{pmatrix}
4 & 5 & 2 \\
1 & -2 & 3
\end{pmatrix} = \begin{pmatrix} 
4 \cdot 4 + 1 \cdot 1 & 4 \cdot 5 + 1 \cdot (-2) & 4 \cdot 2 + 1 \cdot 3\\
5 \cdot 4 + (-2) \cdot 1 & 5 \cdot 5 + (-2) \cdot (-2) &  5 \cdot 2 + (-2) \cdot 3 \\
2 \cdot 4 + 3 \cdot 1 & 2 \cdot 5 + 3 \cdot (-2) & 2 \cdot 2 + 3 \cdot 3
\end{pmatrix} = \begin{pmatrix} 
17 & 18 & 11 \\
18 & 29 & 4 \\
11 & 4 & 13
\end{pmatrix}$$

$$A^{T} \times A =  \begin{pmatrix}
4 & 5 & 2 \\
1 & -2 & 3
\end{pmatrix} \times \begin{pmatrix}
4 & 1\\ 
5 & -2\\ 
2 & 3
\end{pmatrix} = \begin{pmatrix} 
4 \cdot 4 + 5 \cdot 5 + 2 \cdot 2 & 4 \cdot 1 + 5 \cdot (-2) + 2 \cdot 3 \\
1 \cdot 4 + (-2) \cdot 5 + 3 \cdot 2 & 1 \cdot 1 + (-2) \cdot (-2) + 3 \cdot 3
\end{pmatrix} = \begin{pmatrix} 
45 & 0 \\
0 & 14
\end{pmatrix}$$

In [15]:
a = np.array([[4, 1], [5, -2], [2, 3]])

In [16]:
print(f'A x A^T = \n {np.dot(a, a.T)}')

A x A^T = 
 [[17 18 11]
 [18 29  4]
 [11  4 13]]


In [17]:
print(f'A^T x A = \n {np.dot(a.T, a)}')

A^T x A = 
 [[45  0]
 [ 0 14]]


__5*.__ Написать на Python функцию для перемножения двух произвольных матриц, не используя NumPy.

In [18]:
def dot_matrix(a, b):
    if len(a[0]) == len(b) and len(a) == len(b[0]):
        c = [[0 for _ in range(len(a))] for _ in range(len(b[0]))]
        for m in range(len(a)):
            for n in range(len(a[0])):
                for k in range(len(b[0])):
                    c[m][k] = a[m][n] * b[n][k] + c[m][k]
    return c

In [20]:
a = np.array([[4, 3, 8, 5, 4], [8, 2, 0, 4, 2], [9, 1, 2, 5, 9], [7, 3, 6, 8, 7], [2, 3, 1, 4, 9], [4, 1, 8, 2, 0], [4, 6, 2, 2, 2], [9, 8, 7, 5, 6], [4, 7, 9, 3, 4], [1, 3, 5, 3, 0]])
b = np.array([[2, 9, 7, 5, 0, 4, 0, 8, 5, 0], [6, 5, 1, 1, 1, 4, 8, 2, 9, 3], [9, 8, 0, 2, 2, 6, 6, 7, 2, 8], [4, 3, 5, 6, 9, 1, 0, 3, 9, 0], [0, 3, 5, 3, 9, 5, 5, 8, 7, 4]])

In [21]:
a.shape

(10, 5)

In [22]:
b.shape

(5, 10)

In [23]:
np.dot(a, b)

array([[118, 142,  76,  81, 100, 101,  92, 141, 136,  89],
       [ 44, 100,  88,  72,  56,  54,  26,  96, 108,  14],
       [ 62, 144, 134, 107, 131, 102,  65, 175, 166,  55],
       [118, 171, 127, 119, 150, 119,  95, 184, 195,  85],
       [ 47,  80,  82,  66, 122,  75,  75, 113, 138,  53],
       [ 94, 111,  39,  49,  35,  70,  56,  96,  63,  67],
       [ 70,  94,  54,  48,  46,  64,  70,  80, 110,  42],
       [149, 210, 126, 115, 121, 145, 136, 200, 218, 104],
       [143, 164,  70,  75,  88, 121, 130, 150, 156, 109],
       [ 77,  73,  25,  36,  40,  49,  54,  58,  69,  49]])

# Урок 4

# Матрицы и матричные операции. Часть 2

__1.__ Вычислить определитель:

   a)

$$\begin{vmatrix}
sinx & -cosx\\ 
cosx & sinx
\end{vmatrix};$$

   б)
    
$$\begin{vmatrix}
4 & 2 & 3\\ 
0 & 5 & 1\\ 
0 & 0 & 9
\end{vmatrix};$$
    
   в)

$$\begin{vmatrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{vmatrix}.$$

а) $$\begin{vmatrix}
sinx &amp; -cosx\\ 
cosx &amp; sinx
\end{vmatrix} = sinx \cdot sinx - (-cosx) \cdot cosx = 1$$

б) $$\begin{vmatrix}
4 & 2 & 3\\ 
0 & 5 & 1\\ 
0 & 0 & 9
\end{vmatrix} = 4\cdot5\cdot9 = 180$$

In [25]:
a = np.array([[4, 2, 3], [0, 5, 1], [0, 0,9]])
print(f'|A| = {np.linalg.det(a):.0f}')

|A| = 180


в) $$\begin{vmatrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{vmatrix} = 1\cdot5\cdot9 + 7\cdot2\cdot6 + 4\cdot8\cdot3 - 3\cdot5\cdot7 - 1\cdot6\cdot8 - 9\cdot2\cdot4 = 45 + 84 + 96 - 105 - 48 - 72 = 0$$

In [26]:
a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(f'|A| = {np.linalg.det(a):.0f}')

|A| = -0


__2.__ Определитель матрицы $A$ равен $4$. Найти:

   а) $det(A^{2})$;
    
   б) $det(A^{T})$;
    
   в) $det(2A)$.

а) $det(A^{2}) = det(A \cdot A) = det(A) \cdot det(A) = 4 \cdot 4 = 16;$

б) $detA^{T}=detA,$

$detA^{T}=detA=4;$

в) $det(2A) = 2^n \cdot det(A) = 2^n \cdot 4$, где $n$ размер матрицы.

__3.__  Доказать, что матрица

$$\begin{pmatrix}
-2 & 7 & -3\\ 
4 & -14 & 6\\ 
-3 & 7 & 13
\end{pmatrix}$$
   
вырожденная.

$$\begin{vmatrix}
-2 & 7 & -3\\ 
4 & -14 & 6\\ 
-3 & 7 & 13
\end{vmatrix} = (-2)\cdot(-14)\cdot13 + (-3)\cdot7\cdot6 + 4\cdot7\cdot(-3) - (-3)\cdot(-14)\cdot(-3) - (-2)\cdot6\cdot7 - 4\cdot7\cdot13 = 364 - 126 - 84 + 126 + 84 - 364 = 0$$

Если определитель матрицы равен нулю, то матрица вырожденная.

__4.__ Найти ранг матрицы:

   а) $\begin{pmatrix}
1 & 2 & 3\\ 
1 & 1 & 1\\ 
2 & 3 & 4
\end{pmatrix};$

   б) $\begin{pmatrix}
0 & 0 & 2 & 1\\ 
0 & 0 & 2 & 2\\ 
0 & 0 & 4 & 3\\ 
2 & 3 & 5 & 6
\end{pmatrix}.$

а) Третья строка это сумма первой и второй, поэтому её можно отбросить.

$\begin{pmatrix}
1 & 2 & 3\\ 
1 & 1 & 1
\end{pmatrix}$

Теперь вычитаем из первой строки вторую

$\begin{pmatrix}
1 & 2 & 3\\ 
0 & 1 & 2
\end{pmatrix}$

Ранг матрицы 2.

In [31]:
x = [1, 2, 3]
y = [1, 1, 1]
z = [2, 3, 4]
 
a = np.array([x, y, z])
r = np.linalg.matrix_rank(a)
 
print(f'Ранг матрицы: {r}')

Ранг матрицы: 2


б) Третья строка это сумма первых двух, поэтому её можно отбросить.

$\begin{pmatrix}
0 & 0 & 2 & 1\\ 
0 & 0 & 2 & 2\\  
2 & 3 & 5 & 6
\end{pmatrix}$

Вычитаем из второй строки первую.

$\begin{pmatrix}
0 & 0 & 0 & 1\\ 
0 & 0 & 2 & 2\\  
2 & 3 & 5 & 6
\end{pmatrix}$

Ранг матрицы: 3

In [34]:
x = [0, 0, 2, 1]
y = [0, 0, 2, 2]
z = [0, 0, 4, 3]
n = [2, 3, 5, 6]
 
a = np.array([x, y, z, n])
r = np.linalg.matrix_rank(a)
 
print(f'Ранг матрицы: {r}')

Ранг матрицы: 3
